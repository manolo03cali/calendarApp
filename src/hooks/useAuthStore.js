// Importo `useDispatch` para disparar acciones y `useSelector` para acceder al estado desde Redux
import { useDispatch, useSelector } from "react-redux";

// Importo la instancia de mi API personalizada para hacer peticiones HTTP al backend
import { calendarApi } from "../api";

// Importo las acciones del store relacionadas con la autenticación y calendario
import {
  onChecking, // Acción para indicar que estoy verificando las credenciales
  onLogin, // Acción cuando el login fue exitoso
  onLogout, // Acción cuando quiero cerrar sesión o hubo error
  clearErrorMessage, // Limpia mensajes de error
  onLogoutCalendar, // Limpia el estado del calendario al cerrar sesión
} from "../store";

// Este es mi custom hook para manejar toda la lógica de autenticación
export const useAuthStore = () => {
  // Extraigo el estado actual de autenticación desde Redux
  const { status, user, errorMessage } = useSelector((state) => state.auth);

  // Inicializo `dispatch` para poder lanzar acciones
  const dispatch = useDispatch();

  // Esta función la uso para iniciar sesión
  const startLogin = async ({ email, password }) => {
    dispatch(onChecking()); // Cambio el estado a "verificando"

    try {
      // Hago la petición POST al backend para autenticarme
      const {
        data: { token, name, uid },
      } = await calendarApi.post("/auth", { email, password });

      // Si es exitoso, guardo el token y su fecha de inicio en localStorage
      localStorage.setItem("token", token);
      localStorage.setItem("token-init-date", Date.now());

      // Disparo la acción de login con los datos del usuario
      dispatch(onLogin({ name, uid }));
    } catch (error) {
      // Si algo sale mal, disparo logout con un mensaje de error
      dispatch(onLogout("Credenciales incorrectas"));

      // Borro el mensaje de error después de un tiempo muy corto
      setTimeout(() => {
        dispatch(clearErrorMessage());
      }, 10);
    }
  };

  // Esta función la uso para registrar un nuevo usuario
  const startRegister = async ({ name: nombreUsuario, email, password }) => {
    try {
      // Hago una petición POST al backend para crear el usuario
      const {
        data: { token, name, uid },
      } = await calendarApi.post("/auth/new", {
        name: nombreUsuario,
        email,
        password,
      });

      // Guardo token y fecha de inicio
      localStorage.setItem("token", token);
      localStorage.setItem("token-init-date", Date.now());

      // Disparo el login directamente con los datos del usuario nuevo
      dispatch(onLogin({ name, uid }));
    } catch (error) {
      // Si hay error, muestro el mensaje devuelto por el backend
      dispatch(onLogout(error.response.data?.msg));

      // Limpio el mensaje de error rápidamente para que no quede colgado
      setTimeout(() => {
        dispatch(clearErrorMessage());
      }, 10);
    }
  };

  // Esta función la uso para comprobar si el token aún es válido
  const checkAuthToken = async () => {
    const tokenLocal = localStorage.getItem("token"); // Obtengo el token guardado

    if (!tokenLocal) return dispatch(onLogout()); // Si no hay token, cierro sesión

    dispatch(onChecking()); // Cambio estado a verificando

    try {
      // Hago una petición GET al backend para renovar el token
      const {
        data: { name, uid, token },
      } = await calendarApi.get("/auth/renew");

      // Si renueva con éxito, guardo el nuevo token y fecha
      localStorage.setItem("token", token);
      localStorage.setItem("token-init-date", Date.now());

      // Disparo login con los datos del usuario
      dispatch(onLogin({ name, uid }));
    } catch (error) {
      // Si hay error, limpio localStorage y disparo logout
      console.error("Error al renovar token:", error);
      localStorage.clear();
      dispatch(onLogout());
    }
  };

  // Esta función la uso para cerrar sesión completamente
  const startLogout = () => {
    localStorage.removeItem("token"); // Borro el token
    localStorage.removeItem("token-init-date"); // Borro la fecha
    dispatch(onLogoutCalendar()); // Limpio estado del calendario
    dispatch(onLogout()); // Limpio estado de autenticación
  };

  // Devuelvo todas las propiedades y funciones que necesito usar en los componentes
  return {
    // Estado
    errorMessage,
    status,
    user,

    // Métodos
    startLogin,
    startRegister,
    checkAuthToken,
    startLogout,
  };
};
