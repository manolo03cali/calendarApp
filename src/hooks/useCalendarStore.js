// Importo los hooks de Redux para poder interactuar con el store
import { useDispatch, useSelector } from "react-redux";

// Importo las acciones que puedo disparar sobre el slice del calendario
import {
  onSetActiveEvent, // Establece un evento como el activo (seleccionado)
  onAddNewEvent, // Agrega un nuevo evento al store
  onUpdateEvent, // Actualiza un evento existente en el store
  onDeleteEvent, // Elimina el evento activo del store
  onLoadEvents, // Carga los eventos desde la base de datos
} from "../store";

// Importo la API personalizada para hacer peticiones HTTP
import { calendarApi } from "../api";

// Función auxiliar que convierte las fechas de los eventos a objetos Date reales
import { convertEventsToDateEvents } from "../helpers";

// Librería externa para mostrar alertas visuales
import Swal from "sweetalert2";

// Creo un hook personalizado que encapsula toda la lógica relacionada con el calendario
export const useCalendarStore = () => {
  // Obtengo `dispatch` para poder lanzar acciones a Redux
  const dispatch = useDispatch();

  // Extraigo datos del estado global: los eventos, el evento activo y si ya cargué los eventos
  const { events, activeEvent, isLoaded } = useSelector(
    (state) => state.calendar
  );

  // También necesito acceder al usuario autenticado para asociarlo a los eventos
  const { user } = useSelector((state) => state.auth);

  // === Función: seleccionar un evento activo ===
  // Se llama cuando el usuario hace clic en un evento del calendario
  const setActiveEvent = (calendarEvent) => {
    dispatch(onSetActiveEvent(calendarEvent));
  };

  // === Función: guardar un evento (nuevo o existente) ===
  const startSavingEvent = async (calendarEvent) => {
    try {
      // Si el evento ya tiene un id, significa que ya existe → actualizo
      if (calendarEvent.id) {
        await calendarApi.put(`/events/${calendarEvent.id}`, calendarEvent);
        dispatch(onUpdateEvent({ ...calendarEvent, user })); // actualizo con usuario
        return;
      }

      // Si no tiene id, entonces es un evento nuevo → lo creo
      const { data } = await calendarApi.post("/events", calendarEvent);
      dispatch(onAddNewEvent({ ...calendarEvent, id: data.event.id, user })); // le asigno el id del backend
    } catch (error) {
      console.log(error);
      // Si algo falla, muestro una alerta con el mensaje de error del backend
      Swal.fire("Error al guardar", error.response.data.msg, "error");
    }
  };

  // === Función: eliminar el evento actualmente activo ===
  const startDeletingEvent = async () => {
    try {
      await calendarApi.delete(`/events/${activeEvent.id}`); // envío petición DELETE
      dispatch(onDeleteEvent()); // y luego actualizo el store
    } catch (error) {
      console.log(error);
      Swal.fire("Error al eliminar", error.response.data.msg, "error");
    }
  };

  // === Función: cargar los eventos desde el backend ===
  const startLoadingEvents = async () => {
    if (isLoaded) return; // Evito recargar si ya fueron cargados

    console.log("ejecutando startLoadingEvents()");
    try {
      const { data } = await calendarApi.get("/events"); // obtengo los eventos
      const events = convertEventsToDateEvents(data.events); // convierto strings a fechas reales
      dispatch(onLoadEvents(events)); // actualizo el estado con los eventos cargados
    } catch (error) {
      console.log("Error cargando eventos");
      console.log(error);
    }
  };

  // === Retorno de valores y funciones ===
  return {
    // Propiedades del estado del calendario
    events, // Lista de todos los eventos
    activeEvent, // Evento actualmente seleccionado
    hasEventSelected: !!activeEvent, // true si hay evento seleccionado

    // Métodos disponibles para los componentes
    setActiveEvent, // Seleccionar evento
    startSavingEvent, // Guardar o actualizar evento
    startDeletingEvent, // Eliminar evento
    startLoadingEvents, // Cargar eventos
  };
};
