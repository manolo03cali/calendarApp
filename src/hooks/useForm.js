// Importo hooks de React para manejar estado, efectos y memorias
import { useEffect, useMemo, useState } from "react";

// Defino un custom hook que recibe:
// - `initialForm`: el estado inicial del formulario
// - `formValidations`: un objeto con funciones de validación para cada campo
export const useForm = (initialForm = {}, formValidations = {}) => {
  // Este estado almacena los valores actuales del formulario (como el texto de los inputs)
  const [formState, setFormState] = useState(initialForm);

  // Este estado guarda los errores de validación por campo, si existen
  const [formValidation, setFormValidation] = useState({});

  // Este efecto se ejecuta cada vez que cambia el estado del formulario.
  // Llama a la función que evalúa las validaciones.
  useEffect(() => {
    createValidators();
  }, [formState]);

  // Este efecto se ejecuta cuando cambian los valores iniciales del formulario.
  // Reseteo el formulario al nuevo estado inicial.
  useEffect(() => {
    setFormState(initialForm);
  }, [initialForm]);

  // Esta variable calcula si todo el formulario es válido (true) o no (false)
  const isFormValid = useMemo(() => {
    // Recorro todas las validaciones actuales y si alguna tiene error (≠ null), devuelvo false
    for (const formValue of Object.keys(formValidation)) {
      if (formValidation[formValue] !== null) return false;
    }
    // Si todo está correcto (todos son null), devuelvo true
    return true;
  }, [formValidation]);

  // Esta función se ejecuta cada vez que el usuario escribe en un input
  const onInputChange = ({ target }) => {
    const { name, value } = target;
    // Actualizo el estado del formulario con el nuevo valor del input
    setFormState({
      ...formState,
      [name]: value,
    });
  };

  // Esta función permite resetear el formulario a su estado inicial
  const onResetForm = () => {
    setFormState(initialForm);
  };

  // Esta función aplica las validaciones que definí para cada campo
  const createValidators = () => {
    const formCheckedValues = {};

    // Recorro cada campo que tiene validación
    for (const formField of Object.keys(formValidations)) {
      // Extraigo la función de validación y su mensaje de error
      const [fn, errorMessage] = formValidations[formField];

      // Aplico la validación. Si pasa, no hay error (null). Si falla, guardo el mensaje de error.
      formCheckedValues[`${formField}Valid`] = fn(formState[formField])
        ? null
        : errorMessage;
    }

    // Actualizo el estado de validaciones
    setFormValidation(formCheckedValues);
  };

  // Devuelvo lo necesario para que el formulario funcione con este hook
  return {
    ...formState, // Los campos individuales como `email`, `password`, etc.
    formState, // El objeto completo del formulario
    onInputChange, // Función para manejar cambios en los inputs
    onResetForm, // Función para reiniciar el formulario

    ...formValidation, // Propiedades como `emailValid`, `nameValid`, etc.
    isFormValid, // Booleano que dice si todo el formulario es válido
  };
};
