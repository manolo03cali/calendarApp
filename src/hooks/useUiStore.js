// Importo los hooks de React-Redux para interactuar con el store de la app
import { useDispatch, useSelector } from "react-redux";

// Importo las acciones que permiten abrir o cerrar el modal desde el slice de UI
import { onOpenDateModal, onCloseDateModal } from "../store";

// Creo un custom hook llamado `useUiStore` para manejar la lógica relacionada con la UI (interfaz)
export const useUiStore = () => {
  // Obtengo `dispatch`, que me permite enviar acciones al store (como abrir/cerrar el modal)
  const dispatch = useDispatch();

  // Uso `useSelector` para extraer el valor de `isDateModalOpen` desde el estado global (store)
  // Esto me dice si el modal actualmente está abierto (true) o cerrado (false)
  const { isDateModalOpen } = useSelector((state) => state.ui);

  // Esta función abre el modal lanzando la acción `onOpenDateModal`
  const openDateModal = () => {
    dispatch(onOpenDateModal());
  };

  // Esta función cierra el modal lanzando la acción `onCloseDateModal`
  const closeDateModal = () => {
    dispatch(onCloseDateModal());
  };

  // Esta función alterna el estado del modal:
  // - Si está abierto, lo cierra
  // - Si está cerrado, lo abre
  const toggleDateModal = () => {
    isDateModalOpen ? closeDateModal() : openDateModal();
  };

  // Devuelvo las propiedades y funciones que quiero que estén disponibles al usar este hook
  return {
    // Estado del modal (abierto o cerrado)
    isDateModalOpen,

    // Métodos para controlar el modal
    openDateModal,
    closeDateModal,
    toggleDateModal,
  };
};
