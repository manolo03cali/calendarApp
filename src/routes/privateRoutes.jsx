// Primero importo el layout privado (`PrivateLayout`) que se encarga de proteger las rutas.
// Este componente solo deja pasar a los usuarios autenticados.
import { PrivateLayout } from "../layouts";

// Luego importo la página principal del calendario. Esta es la funcionalidad principal
// que quiero proteger para que solo esté disponible cuando el usuario ha iniciado sesión.
import { CalendarPage } from "../calendar";

// También importo una página de error, por si el usuario intenta entrar a una ruta privada que no existe.
import { ErrorPage } from "../auth";

// Ahora defino un arreglo con todas las rutas privadas de la aplicación.
// Estas rutas estarán protegidas por el `PrivateLayout`, que se encargará de redirigir al login
// si el usuario no está autenticado.
export const privateRoutes = [
  {
    // Uso el `PrivateLayout` como componente principal (envoltorio) para todas las rutas privadas.
    element: <PrivateLayout />,

    // Dentro de este layout, defino todas las rutas hijas que quiero proteger.
    children: [
      // Esta ruta se activa cuando el usuario navega a la raíz "/".
      // Gracias a `index: true`, React Router sabe que debe cargar esta ruta por defecto.
      // Aquí muestro la página del calendario, que es la funcionalidad central de mi app.
      { index: true, element: <CalendarPage /> },

      // Esta ruta captura cualquier otra ruta no definida dentro del contexto privado (usando el comodín "*").
      // Si alguien escribe mal una URL o accede a una ruta inexistente, muestro la página de error.
      { path: "*", element: <CalendarPage /> },
    ],
  },
];
