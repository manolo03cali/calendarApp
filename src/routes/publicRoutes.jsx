// Primero importo el `PublicLayout`, que va a envolver todas las rutas que están disponibles
// sin necesidad de estar autenticado (como login o error).
import { PublicLayout } from "../layouts";

// Luego importo las páginas que van dentro de estas rutas públicas:
// - `LoginPage` es la página para iniciar sesión.
// - `ErrorPage` se muestra cuando el usuario intenta acceder a una ruta no válida.
import { LoginPage, ErrorPage } from "../auth";

// Ahora defino un arreglo con todas las rutas públicas de la aplicación.
// Estas rutas serán accesibles incluso si el usuario no ha iniciado sesión.
export const publicRoutes = [
  {
    // Este `element` indica que todas estas rutas estarán envueltas por el `PublicLayout`,
    // el cual se encarga de verificar que el usuario NO esté autenticado.
    element: <PublicLayout />,

    // Dentro de este layout, defino las rutas hijas públicas.
    children: [
      // Esta ruta se activa cuando el usuario va a `/login`.
      // Allí muestro la página para iniciar sesión.
      { path: "login", element: <LoginPage /> },

      // Esta ruta actúa como catch-all: si el usuario intenta acceder a una ruta pública que no existe,
      // le muestro la `ErrorPage`.
      { path: "*", element: <ErrorPage /> },
    ],
  },
];
