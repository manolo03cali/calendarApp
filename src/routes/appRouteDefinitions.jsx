// Primero importo el layout raíz de mi aplicación, que probablemente maneja la lógica de autenticación,
// como verificar si el usuario está logueado, mostrar un loader mientras se valida, y luego permitir el acceso a las rutas.
import { RootLayout } from "../layouts";

// Luego importo las rutas privadas, que están destinadas a usuarios autenticados
// (por ejemplo, dashboard, perfil, etc.).
import { privateRoutes } from "./privateRoutes";

// También importo las rutas públicas, como login o registro, que no requieren autenticación.
import { publicRoutes } from "./publicRoutes";

// Ahora defino todas las rutas principales de la app.
// Le indico que la ruta raíz `/` va a usar el `RootLayout`, que envolverá a todas las rutas hijas.
export const appRouteDefinitions = [
  {
    path: "/", // Esto representa la raíz del sitio
    element: <RootLayout />, // Aquí uso el layout raíz como contenedor principal
    children: [
      // Combino (spread) todas las rutas públicas y privadas bajo este layout.
      // De esta manera, `RootLayout` podrá decidir qué mostrar dependiendo del estado de autenticación.
      ...publicRoutes,
      ...privateRoutes,
    ],
  },
];
