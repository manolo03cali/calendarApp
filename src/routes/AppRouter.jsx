// Primero importo la función `createBrowserRouter` desde React Router.
// Esta función me permite crear un enrutador que se basa en el historial del navegador (como cuando el usuario usa el botón "atrás").
import { createBrowserRouter } from "react-router-dom";

// Luego importo las definiciones de las rutas de mi aplicación.
// Estas rutas las tengo definidas en otro archivo como un arreglo de objetos.
// Cada objeto representa una ruta (con su path, componente, hijos, etc.).
import { appRouteDefinitions } from "./appRouteDefinitions";

// Ahora creo el enrutador principal para mi app usando `createBrowserRouter`.
// Le paso las rutas que importé, y con eso el enrutador ya sabe cómo debe comportarse
// dependiendo de la URL actual (por ejemplo, si estoy en `/login`, muestra el LoginPage).
export const AppRouter = createBrowserRouter(appRouteDefinitions);
