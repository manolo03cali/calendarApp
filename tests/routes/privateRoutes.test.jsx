// Primero importo las funciones necesarias de @testing-library/react para renderizar componentes
// y poder consultar el contenido del DOM dentro de mis pruebas.
import { render, screen } from "@testing-library/react";

// Luego importo herramientas de react-router-dom que me permiten simular rutas en un entorno de pruebas.
// `createMemoryRouter` me sirve para crear un router en memoria y `RouterProvider` para inyectarlo en el renderizado.
import { createMemoryRouter, RouterProvider } from "react-router-dom";

// Importo mi hook personalizado, `useAuthStore`, que normalmente se encarga de decirme si el usuario está autenticado.
import { useAuthStore } from "../../src/hooks/useAuthStore";

// Importo las rutas privadas que definí para la app, las mismas que usan el layout protegido y el calendario.
import { privateRoutes } from "../../src/routes/privateRoutes";

// Hago un mock de mi hook de autenticación para poder controlar el estado de autenticación dentro de los tests.
jest.mock("../../src/hooks/useAuthStore");

// También hago mocks de los componentes de página que uso en las rutas privadas,
// reemplazándolos por versiones simples con texto visible para facilitar las pruebas.
jest.mock("../../src/calendar/pages/CalendarPage", () => ({
  CalendarPage: () => <h1>MockCalendarPage</h1>,
}));
jest.mock("../../src/auth/pages/ErrorPage", () => ({
  ErrorPage: () => <h1>MockErrorPage</h1>,
}));

// Agrupo todos los tests relacionados con rutas privadas en un bloque `describe`.
describe("Pruebas en las rutas privadas", () => {
  // Antes de cada test, me aseguro de limpiar cualquier mock anterior
  // para que no interfiera con el siguiente test.
  beforeEach(() => {
    jest.clearAllMocks();
  });

  // Este primer test verifica que, si el usuario está autenticado, se debe mostrar el `CalendarPage`.
  test("Debe mostrar CalendarPage si el usuario está autenticado", () => {
    // Simulo que el hook de autenticación retorna que el usuario está "authenticated"
    useAuthStore.mockReturnValue({
      status: "authenticated",
    });

    // Creo un router en memoria con las rutas privadas, y le paso como ruta inicial la raíz "/"
    const router = createMemoryRouter(privateRoutes, {
      initialEntries: ["/"], // es como si el usuario entrara a "/"
    });

    // Renderizo el `RouterProvider` con ese router simulado
    const { container } = render(<RouterProvider router={router} />);

    // Verifico que se haya renderizado el texto que puse en el mock del CalendarPage.
    // Si aparece, significa que se cargó correctamente el componente.
    expect(screen.getByText("MockCalendarPage")).toBeInTheDocument();
    expect(container).toMatchSnapshot();
  });

  // Este segundo test verifica que incluso si el usuario entra a una ruta no definida,
  // se debe seguir mostrando el CalendarPage (según mi lógica de SPA centrada en el calendario).
  test("Debe de mostrar CalendarPage en cualquier ruta si está autenticado", () => {
    // Nuevamente simulo que el usuario está autenticado
    useAuthStore.mockReturnValue({
      status: "authenticated",
    });

    // En este caso, simulo que el usuario entra a una ruta no existente como "/cualquier-ruta"
    const router = createMemoryRouter(privateRoutes, {
      initialEntries: ["/cualquier-ruta"],
    });

    // Renderizo el RouterProvider como antes
    render(<RouterProvider router={router} />);

    // Verifico que, a pesar de ser una ruta no definida,
    // el componente `CalendarPage` se sigue mostrando correctamente.
    expect(screen.getByText("MockCalendarPage")).toBeInTheDocument();
  });
});
