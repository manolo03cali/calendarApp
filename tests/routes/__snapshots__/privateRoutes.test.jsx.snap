// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`Pruebas en las rutas privadas Debe mostrar CalendarPage si el usuario est√° autenticado 1`] = `
<div>
  <div
    class="container"
  >
    <h1>
      MockCalendarPage
    </h1>
  </div>
</div>
`;
